File: arcane-sdk-ts/src/resources/evaluations.ts
import type { ArcaneClient } from '../client';
import type {
  Evaluation,
  EvaluationResult,
  ExperimentScores,
  CreateEvaluationParams,
  CreateEvaluationResultParams,
  PendingScoreResult,
  ImportScoreResultRow,
  ImportScoreResultsResponse,
  ListPendingScoreResultsOptions,
} from '../types';

export class EvaluationsResource {
  constructor(private readonly client: ArcaneClient) {}

  async list(): Promise<Evaluation[]> {
    return this.client.get<Evaluation[]>('/api/public/evaluations');
  }

  async get(evaluationId: string): Promise<Evaluation> {
    return this.client.get<Evaluation>(`/api/public/evaluations/${evaluationId}`);
  }

  async getExperimentScores(
    evaluationId: string,
    experimentId: string
  ): Promise<ExperimentScores> {
    return this.client.get<ExperimentScores>(
      `/api/public/evaluations/${evaluationId}/experiments/${experimentId}/scores`
    );
  }

  async create(params: CreateEvaluationParams): Promise<Evaluation> {
    return this.client.post<Evaluation>('/api/public/evaluations', params);
  }

  async createResult(
    evaluationId: string,
    params: CreateEvaluationResultParams
  ): Promise<EvaluationResult> {
    return this.client.post<EvaluationResult>(
      `/api/public/evaluations/${evaluationId}/results`,
      params
    );
  }

  async listPendingScoreResults(
    evaluationId: string,
    scoreId: string,
    options: ListPendingScoreResultsOptions = {}
  ): Promise<{ data: PendingScoreResult[]; pagination: { page: number; limit: number; total: number; totalPages: number; hasNextPage: boolean; hasPreviousPage: boolean } }> {
    const params: Record<string, string | number> = {};
    if (options.experimentId) params.experimentId = options.experimentId;
    if (options.page != null) params.page = options.page;
    if (options.limit != null) params.limit = options.limit;
    return this.client.get(
      `/api/public/evaluations/${evaluationId}/scores/${scoreId}/pending-results`,
      params
    );
  }

  async *listPendingScoreResultsIterator(
    evaluationId: string,
    scoreId: string,
    options: Omit<ListPendingScoreResultsOptions, 'page'> = {}
  ): AsyncGenerator<PendingScoreResult> {
    const limit = options.limit ?? 100;
    let page = 1;
    let hasNext = true;
    while (hasNext) {
      const res = await this.listPendingScoreResults(evaluationId, scoreId, {
        ...options,
        page,
        limit,
      });
      for (const item of res.data) {
        yield item;
      }
      hasNext = res.pagination.hasNextPage;
      page++;
    }
  }

  async importScoreResults(
    evaluationId: string,
    scoreId: string,
    results: ImportScoreResultRow[]
  ): Promise<ImportScoreResultsResponse> {
    return this.client.post<ImportScoreResultsResponse>(
      `/api/public/evaluations/${evaluationId}/scores/${scoreId}/import-results`,
      { results }
    );
  }
}

File: arcane-sdk-py/src/arcane_sdk/resources_async/evaluations.py
"""Evaluations resource (async)."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, AsyncIterator

if TYPE_CHECKING:
    from arcane_sdk.client_async import ArcaneClientAsync


class EvaluationsResourceAsync:
    def __init__(self, client: ArcaneClientAsync) -> None:
        self._client = client

    async def list(self) -> list[dict[str, Any]]:
        """List all evaluations."""
        return await self._client._get("/api/public/evaluations")

    async def get(self, evaluation_id: str) -> dict[str, Any]:
        """Get an evaluation by ID."""
        return await self._client._get(
            f"/api/public/evaluations/{evaluation_id}"
        )

    async def get_experiment_scores(
        self,
        evaluation_id: str,
        experiment_id: str,
    ) -> dict[str, Any]:
        """Get experiment scores for an evaluation."""
        return await self._client._get(
            f"/api/public/evaluations/{evaluation_id}/experiments/{experiment_id}/scores"
        )

    async def create(self, params: dict[str, Any]) -> dict[str, Any]:
        """Create an evaluation."""
        return await self._client._post(
            "/api/public/evaluations",
            json=params,
        )

    async def create_result(
        self,
        evaluation_id: str,
        params: dict[str, Any],
    ) -> dict[str, Any]:
        """Create an evaluation result (manual scoring)."""
        return await self._client._post(
            f"/api/public/evaluations/{evaluation_id}/results",
            json=params,
        )

    async def list_pending_score_results(
        self,
        evaluation_id: str,
        score_id: str,
        *,
        experiment_id: str | None = None,
        page: int = 1,
        limit: int = 100,
    ) -> dict[str, Any]:
        """
        List pending score results for a manual score (status PENDING).
        For dataset-scoped evaluations: omit experiment_id.
        For experiment-scoped evaluations: pass experiment_id.
        Returns paginated results.
        """
        params: dict[str, Any] = {"page": page, "limit": limit}
        if experiment_id is not None:
            params["experimentId"] = experiment_id
        return await self._client._get(
            f"/api/public/evaluations/{evaluation_id}/scores/{score_id}/pending-results",
            params=params,
        )

    async def list_pending_score_results_iterator(
        self,
        evaluation_id: str,
        score_id: str,
        *,
        experiment_id: str | None = None,
        limit: int = 100,
    ) -> AsyncIterator[dict[str, Any]]:
        """
        Iterate over all pending score results for a manual score.
        For dataset scope: omit experiment_id.
        For experiment scope: pass experiment_id.
        """
        page = 1
        while True:
            res = await self.list_pending_score_results(
                evaluation_id,
                score_id,
                experiment_id=experiment_id,
                page=page,
                limit=limit,
            )
            for item in res.get("data", []):
                yield item
            pagination = res.get("pagination", {})
            if not pagination.get("hasNextPage", False):
                break
            page += 1

    async def import_score_results(
        self,
        evaluation_id: str,
        score_id: str,
        results: list[dict[str, Any]],
    ) -> dict[str, Any]:
        """
        Import/update score results (bulk).
        For dataset scope: use datasetRowId per row.
        For experiment scope: use experimentResultId per row.
        Upserts existing PENDING results or creates new ones.
        Each row: {"datasetRowId"?: str, "experimentResultId"?: str, "value": number, "reasoning"?: str}
        """
        return await self._client._post(
            f"/api/public/evaluations/{evaluation_id}/scores/{score_id}/import-results",
            json={"results": results},
        )
